verification/cbmc/sources/utils.c

```diff
@@ -0,0 +1,19 @@
+#include <utils.h>
+
+void assert_byte_buf_contents_are_equal(const struct aws_byte_buf *const lhs, const struct aws_byte_buf *const rhs) {
+    assert(lhs && rhs);
+    assert(lhs->len == rhs->len);
+    if (lhs->len > 0) {
+        size_t index;
+        __CPROVER_assume(index < lhs->len);
+        assert(lhs->buffer[index] == rhs->buffer[index]);
+    }
+}
+
+void assert_keys_are_equal(const struct content_key *ckey, const struct data_key *dkey, const size_t max_len) {
+    assert(ckey && dkey);
+    assert(ckey->keybuf && dkey->keybuf);
+    size_t index;
+    __CPROVER_assume(index < max_len);
+    assert(ckey->keybuf[index] == dkey->keybuf[index]);
+}
```

Comment:
If they are both null, they are both equal. You don't need this `max_len`, because these data structures internally have known sizes, i.e., `MAX_DATA_KEY_SIZE`. So you could just write.
```c
void assert_keys_are_equal(const struct content_key *ckey, const struct data_key *dkey) {
    assert(ckey && dkey);
    assert(ckey->keybuf && dkey->keybuf);
    size_t index;
    __CPROVER_assume(index < MAX_DATA_KEY_SIZE);
    assert(ckey->keybuf[index] == dkey->keybuf[index]);
}
```

Reply:
I was using that before but the proof fails because this is what happens in derive_key_v1:
```
memcpy(content_key->keybuf, data_key->keybuf, props->data_key_len);
```
and previously the content_key is zero'd:
```
aws_secure_zero(content_key->keybuf, sizeof(content_key->keybuf));
```
So I think we should keep it as it is (except for the case where both are NULL).